tardis/simulation/base.py

```diff
@@ -401,6 +406,146 @@ def to_hdf(self, model, path_or_buf, path='', plasma_properties=None):
         self.runner.to_hdf(path_or_buf, path)
         model.to_hdf(path_or_buf, path, plasma_properties)
 
+    def make_source_function(self, model):
+        """
+        Calculates the source function using the line absorption rate estimator `Edotlu_estimator`
+
+        Formally it calculates the expresion ( 1 - exp(-tau_ul) ) S_ul but this product is what we need later,
+        so there is no need to factor out the source function explicitly.
+
+        Parameters
+        ----------
+        model : tardis.model.Radial1DModel
+
+        Returns
+        -------
+        Numpy array containing ( 1 - exp(-tau_ul) ) S_ul ordered by wavelength of the transition u -> l
+        """
+
+        upper_level_index = model.atom_data.lines.set_index(['atomic_number', 'ion_number', 'level_number_upper']).index.copy()
+        e_dot_lu          = pd.DataFrame(self.runner.Edotlu, index=upper_level_index)
+        e_dot_u           = e_dot_lu.groupby(level=[0, 1, 2]).sum()
+        e_dot_u.index.names = ['atomic_number', 'ion_number', 'source_level_number'] # To make the q_ul e_dot_u product work, could be cleaner
+        transitions       = model.atom_data.macro_atom_data[model.atom_data.macro_atom_data.transition_type == -1].copy()
+        transitions_index = transitions.set_index(['atomic_number', 'ion_number', 'source_level_number']).index.copy()
+        tmp  = model.plasma.transition_probabilities[(model.atom_data.macro_atom_data.transition_type == -1).values]
+        q_ul = tmp.set_index(transitions_index)
+        t    = model.tardis_config.supernova.time_explosion.value
+        wave = model.atom_data.lines.wavelength_cm[transitions.transition_line_id].values.reshape(-1,1)
+        reorder  = wave[:,0].argsort()
+        att_S_ul =  ( wave * (q_ul * e_dot_u) * t  / (4*np.pi) ).iloc[reorder,:]
+
```

Comment:
Not sure that everything goes as planned here. Maybe use a more simplistic approach for testing purposes:
```python
        line_ids = model.atom_data.lines.index
        upper_levels = model.atom_data.lines.level_number_upper.values
        lower_levels = model.atom_data.lines.level_number_lower.values
        t = model.tardis_config.supernova.time_explosion

        Edot_u = {}
        j_ul = {}
        S_ul = {}

        for i in range(1, int(max(upper_levels)+1), 1):
            Edot_u[i] = model.runner.Edotlu[upper_levels == i].sum()

        for i in xrange(len(line_ids)):
            line_id = line_ids[i]
            j_ul[line_id] = 1. / (4. * np.pi) * model.plasma.transition_probabilities.ix[line_id].values * Edot_u[iup[i]]
            S_ul[line_id] = (j_ul[line_id] * t * u.AA * model.atom_data.lines.ix[line_id].wavelength).to("1/cm^2")
```